lib/rspec/rails/matchers/have_reported_error.rb

```diff
@@ -0,0 +1,272 @@
+require 'active_support/testing/error_reporter_assertions'
+
+module RSpec
+  module Rails
+    module Matchers
+      ErrorCollector = ActiveSupport::Testing::ErrorReporterAssertions::ErrorCollector
+
+      # Matcher class for `have_reported_error`. Should not be instantiated directly.
+      #
+      # Provides a way to test that an error was reported to Rails.error.
+      #
+      # @api private
+      # @see RSpec::Rails::Matchers#have_reported_error
+      class HaveReportedError < RSpec::Rails::Matchers::BaseMatcher
+        # @param expected_error_or_message [Class, String, Regexp, nil]
+        #   Error class, message string, or message pattern
+        # @param expected_message [String, Regexp, nil]
+        #   Expected message when first param is a class
+        def initialize(expected_error_or_message, expected_message)
+          @attributes = {}
+          @warn_about_nil_error = expected_error_or_message.nil?
+
+          case expected_error_or_message
+          when UNDEFINED
+            @expected_error = nil
+            @expected_message = expected_message
```

Comment:
If the first argument was not passed at all and defaulted to `UNDEFINED`, the second, `expected_message` will also be `nil`?
```suggestion
            @expected_error = @expected_message = nil
```